### ISSUE TITLE
Position based renderEach for Entities without ID

### ISSUE THREAD

**Issue by metin-kale-cf** [jwstegemann/fritz2#645]
When using renderEach we have to provide an IDProvider:

     store.renderEach(Entity::id) {}

But when using entities without IDs renderEach is not properly working.

Example:

    val store : Store<List<Entity>> = ...
    
    store.renderEach ({ it } /*no unique identifier in Entity, we have to use element itself here*/) { substore -> 
        inputField(value = substore.sub(Entity.value)){}
    }

Initial rendering works fine, but when updating the store, the lens used in renderEach compares the entities using the IDProvider and can't match the correct item of the element. The change will be discarded.

Following code works, because the element is tied with the index:

    store.data.map{ it.size }.distinctUntilChanged().render {
        ( 0 until it ).map { store.sub(index = it) }.forEach(){ substore ->
            inputField(value = substore.sub(Entity.value)){}
        }
    }

As solution we should be able to set a Index-Based ID-Provider.
For example we could pass the index as secondary parameter to the IDProvider:

     store.renderEach({ _, index -> index }) {}

Or the IDProvider could use IndexedValue<T>, so we can use:

     store.renderEach({it.value}) {}

Or analog to kotlins forEachIndexed we could add a renderEachIndexed, without touching the existing api

     store.renderEachIndexed( { _, index -> index } ) { store, index -> } 

As shown above, there is actually a workaround, but it would be more convenient if it would be possible to do it with renderEach.

**Comment by chausknecht** [comment]
I do not think that this would a a good idea, as you simply cannot provide some `renderEach` functionality without a stable identification! An index on lists is definitely not stable at all - insert an item, delete some item or sort the collection will break this so easily.

Your example works, as it is simply no *precise* data-binding! You just render the whole collection every time its size changes (which btw will fail if someone replaces items for example by sorting or alike). A proper solution would be to simply use `render` on the whole collection without any "magical optimization"  trick and to render the structure with `forEachIndexed`.

The `renderEach`-variants are meant for supporting update functionality for collections with very efficient (often minimal) DOM manipulations - aka *precise* data-binding. This intention inherently needs the possibility to identifiy an element from the new state within those from the old (already rendered) ones. There is no way to break out - if you need this, you simply have to adapt your model to support some stable identifier!

What I have encountered while thinking about this problem is the following shortcome of our sub-store mechanics:
```kotlin
fun main() {
    val frameworkStore = storeOf(listOf(Framework("fritz2"), Framework("React"), Framework("Spring")))

    render {
        button {
            label { +"Changes" }
            clicks.map {
                frameworkStore.current.map { if (it.name == "React") Framework("Redux") else it }
            } handledBy frameworkStore.update
        }
        ul {
            frameworkStore.data.renderEach({ it }) { framework ->
                console.log("rendere $framework")
                val storedFramework = frameworkStore.sub(framework) { it }
                //                                                  ^^^^^^
                // This absolutely must "throw" (log) an error at runtime!

                val storedName = storedFramework.sub(Framework.name())
                li {
                    p { +framework.name }
                    p { storedName.data.renderText() }
                    input {
                        value(storedName.data)
                        changes.values() handledBy storedName.update
                    }
                }
            }
        }
    }
}
```
We should definitly think about throwing some error, if there is no matching in the `RootStore` defined by the `idProvider` of the sub-store: As the `idProvider` is not stable, chaging the value will lead to the situation, that the store cannot find any corresponding element. This imho clearly shows an error. 

Currently fritz2 silently "swallows" the updated data and will not propagate it upwards.

**Comment by chausknecht** [comment]
Detected misbehaviour is addressed by issue #662.

@metin-kale-cf Do you have any further reasons to left this issue open? Or are you d'accord with my claims?
I would appreciate to close this in the latter case.
